community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,6 +75,52 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  /**
+   * Finds all relationships connecting fromNode and toNode.
+   */
+  private def findRelationships(query: QueryContext, fromNode: Node, toNode: Node, relCache: RelationshipsCache) = {
+    //check degree and iterate from the node with smaller degree
+    val relTypes = types.types(query)
+    val fromDegree = getDegree(fromNode, relTypes, dir, query)
+    val toDegree = getDegree(toNode, relTypes, dir.reverse, query)
+
+    val (start, end, relationships) = if (fromDegree < toDegree)
+      (fromNode, toNode, query.getRelationshipsForIds(fromNode, dir, relTypes))
+    else
+      (toNode, fromNode, query.getRelationshipsForIds(toNode, dir.reverse(), relTypes))
+
+    if (toNode == null)
+      Iterator.empty
+    else {
+      val size = if (fromDegree < toDegree) fromDegree else toDegree
+      new PrefetchingIterator[Relationship] {
+        val connectedRelationships = new ArrayBuffer[Relationship](size)
+
+        override def fetchNextOrNull(): Relationship = {
+          while (relationships.hasNext) {
+            val rel: Relationship = relationships.next()
+            val other = rel.getOtherNode(start)
```

Comment:
I think I am missing something here. In order to know if we should pick the start node or the end node we need to look at the direction right? From what I can see `rel.getOtherNode` just does one (or ok two sometimes) compares to find the node for me, I cannot see how I can do much better than that that here?